File: python/middle_of_the_linked_list.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def middleNode(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        target = int(self.getDepth(head, 1) / 2)
        return self.getTarget(head, target, 1)

    def getDepth(self, node, count):
        """
        :type node: ListNode
        :type count: int
        :rtype: int
        """
        if node.next == None:
            return count
        
        return self.getDepth(node.next, count + 1)

    def getTarget(self, node, target, current):
        """
        :type node: ListNode
        :type target: int
        :type current: int
        :rtype: ListNode
        """
        if current > target:
            return node

        return self.getTarget(node.next, target, current + 1)
        
      

File: java/src/main/java/com/example/leetcode/MiddleOfTheLinkedList.java
package com.example.leetcode;


public class MiddleOfTheLinkedList {

    ListNode middleNode(ListNode head) {
        int target = getDepth(head) / 2;
        return getTarget(head, target);
    }

    private ListNode getTarget(ListNode node, int target) {
        return getTarget(node, target, 1);
    }

    private ListNode getTarget(ListNode node, int target, int current) {
        if (current > target) {
            return node;
        }

        return getTarget(node.next, target, ++current);
    }

    private int getDepth(ListNode node) {
        return getDepth(node, 1);
    }

    private int getDepth(ListNode node, int depth) {
        if (node.next == null) {
            return depth;
        }

        return getDepth(node.next, ++depth);
    }

    /**
     * Definition for singly-linked list.
     */
    static class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }
}
